File: src/generate-brief.ts
/**
 * generate-brief.ts
 *
 * Reads today's data.json, builds a prompt embedding the full daily-prep
 * skill logic, then invokes `claude` CLI (Claude Code) in --print mode
 * to generate the brief. Output is written to DATA_DIR/YYYY-MM-DD/brief.md
 *
 * Claude Code CLI flags used:
 *   --print                       non-interactive, print response to stdout then exit
 *   --dangerously-skip-permissions  suppress permission prompts for automation
 *   -p "<prompt>"                 the prompt to run
 */

import * as fs from "fs";
import * as path from "path";
import { spawnSync } from "child_process";
import * as dotenv from "dotenv";

dotenv.config({ path: path.resolve(__dirname, "../.env") });

const DATA_DIR = process.env.DATA_DIR || path.resolve(__dirname, "../brief_output");
const TIMEZONE = process.env.TIMEZONE || "America/Chicago";

function toLocalDateString(date: Date, tz: string): string {
  return date.toLocaleDateString("en-CA", { timeZone: tz });
}

function buildPrompt(dataJson: string, today: string): string {
  // Embed the raw data JSON directly in the prompt. Claude Code will
  // reason over it without needing MCP tool calls — the data is already
  // fetched by fetch-data.ts.
  return `You are a sharp chief of staff running Ben's daily planning workflow.

Today's date is ${today}. Timezone: ${TIMEZONE}.

Below is the pre-fetched data for today: Todoist tasks and Google Calendar events.
Do NOT call any external tools — all data you need is here.

<data>
${dataJson}
</data>

<instructions>
Produce Ben's daily brief using the following rules exactly.

## Priority mapping
Todoist's API returns priority as an integer where 4=P1 (urgent), 3=P2, 2=P3, 1=normal/P4.
Always translate before display: API priority 4 → P1, 3 → P2, 2 → P3, 1 → P4.

## Projects
The \`todoist.projects\` map gives project id → name. Match each task's project_id to get its name.
Tasks in the project named "Redox" are Work. Everything else is Personal.

## Labels
Each task's \`labels\` array contains label names directly (e.g. "30min", "important"). Use them as-is.

## Step 1: Understand the data
- \`todoist.todayAndOverdue\` — tasks due today OR overdue (due_date < today)
- \`todoist.importantUndated\` — @important tasks not due today/overdue (may have future due dates)
- \`calendar.today\` — all events today (00:00–23:59)
- \`calendar.tomorrowAM\` — events tomorrow 6 AM–11 AM (for prep-gap detection)

## Step 2: Bucket and estimate
Work = Redox project. Personal = everything else.

Time estimates — apply in order:
1. Labels: 15min / 30min / 1hour → use literally
2. Heuristic by verb:
   - submit / send / file / ping / repull → 15m
   - review / scope / check / pull / read → 30m
   - create / draft / build / write / work on → 60m
   - plan / put together / one-pager / proposal → 60–90m
3. Truly ambiguous → 30m, flagged as ~30m?

Overdue = due.date < today → mark [overdue]
Stale deadline = deadline.date < today → mark [deadline expired DATE]

## Step 3: Compute the day
Working window: 9 AM–5 PM Central = 480 min.
Meeting minutes = sum duration of opaque calendar events (transparency != "transparent") overlapping 9–5.
Free minutes = 480 − meeting minutes.
If meeting minutes > 240, subtract another 30 min (context-switch tax).

Deep-work block: longest contiguous gap between 6 AM and noon, counting:
- Pre-9 AM transparent/self-scheduled blocks (Focus Time, etc.)
- 9 AM onward gaps with no opaque event

## Step 4: Surface flags
- Overcommitment: estimates > 1.25× free time → flag with numbers + drop list
- Calendar/Todoist drift: for each all-day event, check if a task matches by shared proper nouns/acronyms
- Stale deadlines
- Tomorrow-AM prep gaps: external attendees, customer names, agenda gaps + Ben is speaker
- Personal anchors: timed personal events (school pickup, appointments)

## Step 5: Pick morning deep-work item
Walk tiebreakers:
1. Calendar-linked P1 with external dependency due today
2. P1 that fits the block size
3. P1 + @important
4. Any remaining P1
5. P2 + @important (only if no P1s)
6. Any @important
Justify in one sentence.

## Step 6: Output format
Use this exact structure:

\`\`\`
## Daily Prep — [Day, Date]

**[One-line: meeting load, free time, verdict.]**

### Top of mind
- [flags — omit section only if everything is clean]

### Morning deep-work pick
**[Task]** — [why, one sentence]
Block: [HH:MM–HH:MM] ([N] min)

### Work — ranked
1. **[Task]** [P1] [@important] (~30m) [overdue|deadline expired if applicable]
2. ...

### Personal — ranked
1. **[Task]** (~15m)
2. ...

### Today's calendar
- HH:MM–HH:MM  [Event]  [flags]
- ...

### Suggested timeline
- HH:MM–HH:MM  Deep work: [pick]
- HH:MM–HH:MM  [Meeting or Task]
- ...
- 5:00  Wrap

### Undated @important (not slotted)
- [Task] — slot today? (y/n)
\`\`\`

## Tone
Direct. No fluff. Short lines, bold names, skim-friendly.
Push back on overcommitment with numbers, not vibes.
If the day is light, say so.
No emojis anywhere. This prints on paper in a monospaced font.

Do not add any preamble or postamble outside the brief format above.
</instructions>`;
}

async function main() {
  const now = new Date();
  const today = toLocalDateString(now, TIMEZONE);
  const dir = path.join(DATA_DIR, today);
  const dataPath = path.join(dir, "data.json");
  const briefPath = path.join(dir, "brief.md");

  if (!fs.existsSync(dataPath)) {
    console.error(`[generate-brief] No data file found at ${dataPath}`);
    console.error(`[generate-brief] Run fetch-data.ts first.`);
    process.exit(1);
  }

  const dataJson = fs.readFileSync(dataPath, "utf8");
  const prompt = buildPrompt(dataJson, today);

  // Write prompt for debugging
  const promptPath = path.join(dir, "prompt.txt");
  fs.writeFileSync(promptPath, prompt);

  console.log(`[generate-brief] Invoking Claude Code for ${today}...`);

  const result = spawnSync(
    "claude",
    ["--print", "--dangerously-skip-permissions", "--model", "claude-sonnet-4-6", "-p", prompt],
    {
      encoding: "utf8",
      timeout: 300_000, // 5 minutes
      maxBuffer: 10 * 1024 * 1024,
      stdio: ["ignore", "pipe", "pipe"],
    }
  );

  if (result.status !== 0 || result.error) {
    console.error("[generate-brief] Claude Code failed:");
    console.error("  exit code:", result.status);
    console.error("  signal:", result.signal);
    console.error("  error:", result.error?.message);
    console.error("  stderr:", result.stderr || "(empty)");
    console.error("  stdout:", result.stdout?.slice(0, 500) || "(empty)");
    process.exit(1);
  }

  let brief: string = result.stdout;

  // Clean any residual ANSI codes just in case
  brief = brief.replace(/\x1B\[[0-9;]*[mGKHF]/g, "").trim();

  fs.writeFileSync(briefPath, brief);
  console.log(`[generate-brief] ✓ Brief written to ${briefPath}`);
  console.log(`[generate-brief]   ${brief.split("\n").length} lines`);
}

main().catch((err) => {
  console.error("[generate-brief] ERROR:", err.message);
  process.exit(1);
});

File: generate_brief.py
#!/usr/bin/env python3
"""
generate_brief.py

Reads today's data.json, builds the work-task list, personal-task list, and
calendar timeline directly from the data (deterministic — identical structure
every day), then calls Claude via the Amazon Bedrock Converse API for a short
recommendations note. Combines both into the final brief.

Output: DATA_DIR/YYYY-MM-DD/brief.md
"""

import json
import os
import sys
from datetime import date, datetime
from pathlib import Path
from zoneinfo import ZoneInfo

import boto3
from botocore.config import Config
from botocore.exceptions import BotoCoreError, ClientError
from dotenv import load_dotenv

load_dotenv()

DATA_DIR = Path(os.environ.get("DATA_DIR", Path(__file__).parent / "brief_output"))
TIMEZONE = os.environ.get("TIMEZONE", "America/Chicago")

# Bedrock config — BEDROCK_MODEL_ID uses the US cross-region inference profile
# ("us." prefix) by default; override to experiment with other models.
AWS_REGION = os.environ.get("AWS_REGION", "us-east-1")
BEDROCK_MODEL_ID = os.environ.get("BEDROCK_MODEL_ID", "us.anthropic.claude-sonnet-4-6")
MAX_TOKENS = int(os.environ.get("BEDROCK_MAX_TOKENS", "1024"))

PRIORITY_LABELS = {4: "P1", 3: "P2", 2: "P3", 1: "P4"}

# Time-estimate heuristic, applied in order: a matching label wins outright,
# otherwise the task's leading verb picks a bucket, otherwise ~30m.
LABEL_ESTIMATES = {"15min": 15, "30min": 30, "1hour": 60}
VERB_ESTIMATES = (
    (15, ("submit", "send", "file", "ping", "repull")),
    (30, ("review", "scope", "check", "pull", "read")),
    (60, ("create", "draft", "build", "write", "work on")),
    (75, ("plan", "put together", "one-pager", "proposal")),
)


def estimate_minutes(task: dict) -> int:
    for label in task.get("labels") or []:
        if label in LABEL_ESTIMATES:
            return LABEL_ESTIMATES[label]
    content = (task.get("content") or "").strip().lower()
    for minutes, verbs in VERB_ESTIMATES:
        if any(content.startswith(verb) for verb in verbs):
            return minutes
    return 30


def format_task_line(task: dict, today: str) -> str:
    flags = []

    priority_label = PRIORITY_LABELS.get(task.get("priority"), "P4")
    if priority_label != "P4":
        flags.append(f"[{priority_label}]")
    if "important" in (task.get("labels") or []):
        flags.append("[@important]")
    flags.append(f"(~{estimate_minutes(task)}m)")

    due = task.get("due") or {}
    if due.get("date") and due["date"] < today:
        flags.append("[overdue]")

    deadline = task.get("deadline") or {}
    if deadline.get("date") and deadline["date"] < today:
        days_ago = (date.fromisoformat(today) - date.fromisoformat(deadline["date"])).days
        flags.append(f"[deadline expired {deadline['date']} ({days_ago} days ago)]")

    return f"- **{task['content']}** " + " ".join(flags)


def render_task_list(title: str, tasks: list[dict], today: str) -> str:
    lines = [f"### {title}"]
    if not tasks:
        lines.append("- Nothing scheduled.")
    else:
        ranked = sorted(tasks, key=lambda t: (-(t.get("priority") or 1), t.get("content") or ""))
        lines.extend(format_task_line(t, today) for t in ranked)
    return "\n".join(lines)


def format_event_line(event: dict, tz: ZoneInfo) -> str:
    summary = event.get("summary", "(untitled)")
    location = event.get("location", "")
    start = event.get("start", {})
    end = event.get("end", {})

    if "dateTime" in start:
        start_dt = datetime.fromisoformat(start["dateTime"]).astimezone(tz)
        end_dt = datetime.fromisoformat(end["dateTime"]).astimezone(tz)
        line = f"- {start_dt:%H:%M}–{end_dt:%H:%M}  {summary}"
    else:
        line = f"- All day       {summary}"

    if location:
        line += f"  ({location})"
    return line


def render_timeline(events: list[dict], tz: ZoneInfo) -> str:
    lines = ["### Today's schedule"]
    if not events:
        lines.append("- Nothing on the calendar.")
    else:
        lines.extend(format_event_line(e, tz) for e in events)
    return "\n".join(lines)


def build_recommendations_prompt(data_json: str, today: str, weekday: str, sections: str) -> str:
    return f"""You are Ben's executive assistant writing the recommendations note at the bottom of his daily brief.

Today is {today} ({weekday}). Timezone: {TIMEZONE}.

The work-task list, personal-task list, and calendar timeline below are FINAL — they were already
built and printed elsewhere in the brief. Do not repeat them, reformat them, or contradict them.

<finalized-sections>
{sections}
</finalized-sections>

Here is the full underlying data, for numbers and details not shown above (postponed_count,
deadline dates, tomorrow's morning calendar, transparency, etc.):

<data>
{data_json}
</data>

<instructions>
Write ONLY a short, direct recommendations note — at most 5 bullets. Skip anything that doesn't
genuinely apply; a light day should produce a short note, not padding. Never invent a number —
only cite figures that actually appear in the data.

Consider, in rough priority order:
- Overcommitment: if the listed tasks' estimated minutes clearly exceed the free time between
  today's opaque (transparency != "transparent") calendar events, say so with the numbers and name
  1-2 drop candidates (lowest priority / largest estimate, both already in the lists above).
- What to tackle first, and why — one sentence, pointing at a specific task already in the lists above.
- Stale deadlines: any task where deadline.date is before {today} — name it with the days-ago count.
- Chronic postponers: any task with postponed_count >= 10 — name it with the count.
- Tomorrow-AM prep gap: a calendar.tomorrowAM event where Ben is presenting or external/customer
  attendees are present, and no matching prep task exists in today's lists — flag it in one line.

Output format — exactly this, nothing else:

### Recommendations
- ...
- ...

Tone: blunt, specific, skim-friendly. No questions, no preamble, no closing remarks, no emojis.
</instructions>"""


def main() -> None:
    tz = ZoneInfo(TIMEZONE)
    now = datetime.now(tz)
    today = now.strftime("%Y-%m-%d")
    weekday = now.strftime("%A")

    out_dir = DATA_DIR / today
    data_path = out_dir / "data.json"
    brief_path = out_dir / "brief.md"

    if not data_path.exists():
        print(f"[generate-brief] No data file found at {data_path}", file=sys.stderr)
        print("[generate-brief] Run fetch_data.py first.", file=sys.stderr)
        sys.exit(1)

    data_json = data_path.read_text()
    data = json.loads(data_json)

    todoist = data["todoist"]
    work_ids = set(todoist.get("workProjectIds") or [])
    tasks = todoist.get("todayAndOverdue") or []
    work_tasks = [t for t in tasks if t.get("project_id") in work_ids]
    personal_tasks = [t for t in tasks if t.get("project_id") not in work_ids]

    work_section = render_task_list("Work tasks", work_tasks, today)
    personal_section = render_task_list("Personal tasks", personal_tasks, today)
    timeline_section = render_timeline(data["calendar"].get("today") or [], tz)
    sections = "\n\n".join([work_section, personal_section, timeline_section])

    prompt = build_recommendations_prompt(data_json, today, weekday, sections)
    prompt_path = out_dir / "prompt.txt"
    prompt_path.write_text(prompt)

    print(f"[generate-brief] Invoking Bedrock ({BEDROCK_MODEL_ID}) for {today}...")

    client = boto3.client(
        "bedrock-runtime",
        region_name=AWS_REGION,
        config=Config(read_timeout=300, retries={"max_attempts": 3}),
    )

    try:
        resp = client.converse(
            modelId=BEDROCK_MODEL_ID,
            messages=[{"role": "user", "content": [{"text": prompt}]}],
            inferenceConfig={"maxTokens": MAX_TOKENS},
        )
    except (BotoCoreError, ClientError) as e:
        print(f"[generate-brief] ERROR: Bedrock call failed: {e}", file=sys.stderr)
        sys.exit(1)

    recommendations = resp["output"]["message"]["content"][0]["text"].strip()

    header = f"## Daily Prep — {weekday}, {now.strftime('%B %-d')}"
    brief = "\n\n".join([header, work_section, personal_section, timeline_section, recommendations])

    brief_path.write_text(brief)
    print(f"[generate-brief] ✓ Brief written to {brief_path}")
    print(f"[generate-brief]   {len(brief.splitlines())} lines")


if __name__ == "__main__":
    try:
        main()
    except Exception as e:
        print(f"[generate-brief] ERROR: {e}", file=sys.stderr)
        sys.exit(1)
